libsel4/tools/generator_rust.py

```diff
@@ -0,0 +1,604 @@
+        }
+        if name in type_trans:
+            return type_trans[name]
+        else:
+            return name
+
+    @staticmethod
+    def translate_expr(name):
+        # TODO there are cases which implement this in-line, refactor to use this function instead
+        #      or get rid of this
+        if name == "type":
+            return "type_"
+        return name
+
+    TYPES = {
+        8:  "u8",
+        16: "u16",
+        32: "u32",
+        64: "u64"
+    }
+
+    class Type(object):
+        """
+        This class represents a Rust type (such as an 'u32', structure or
+        pointer.
+        """
+
+        def __init__(self, name, size_bits, wordsize, *, native_size_bits=None):
+            """
+            Define a new type, named 'name' that is 'size_bits' bits
+            long.
+            """
+
+            self.name = name
+            self.size_bits = size_bits
+            self.wordsize = wordsize
+            self.double_word = (size_bits == wordsize*2)
+
+            #
+            # Store the number of bits Rust will use for this type
+            # in its native unpacked form.
+            if native_size_bits:
+                self.native_size_bits = native_size_bits
+            else:
+                self.native_size_bits = size_bits
+
+        def pass_by_reference(self):
+            return self.size_bits > self.wordsize and not self.double_word
+
+        def render_parameter_name(self, name):
+            """
+            Return a string of Rust code that would be used in a function
+            parameter declaration.
+            """
+            if name == "type":
+                name = "type_"
+            return "%s: %s" % (name, Generator_Rust.translate_type(self.name))
+
+        def pointer(self, mutable):
+            """
+            Return a new Type class representing a pointer to this object.
+            """
+            return Generator_Rust.PointerType(self, self.wordsize, mutable)
+
+        def rust_expression(self, var_name, word_num=0):
+            """
+            Return code for an expression that gets word 'word_num'
+            of this type.
+            """
+            assert word_num == 0
+            return "%s" % var_name
+
+        def double_word_expression(self, var_name, word_num, word_size):
+
+            assert word_num == 0 or word_num == 1
+
+            if word_num == 0:
+                return "{1} as {0}".format(Generator_Rust.TYPES[self.size_bits], var_name)
+            elif word_num == 1:
+                return "{1}.wrapping_shr({2}) as {0}".format(Generator_Rust.TYPES[self.size_bits], var_name,
+                                                             word_size)
+
+    class BitFieldType(Type):
+        """
+        A special 'struct' generated by the bitfield generator
+        """
+
+        def __init__(self, name, size_bits, wordsize):
+            Generator_Rust.Type.__init__(self, name, size_bits, wordsize)
```

Comment:
Pin Python3 we can use the generic `super().__init__(self, name, size_bits, wordsize)`